presto-main/src/main/java/com/facebook/presto/execution/QueryQueue.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.concurrent.AsyncSemaphore;
+import io.airlift.concurrent.SetThreadName;
+import org.weakref.jmx.Managed;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class QueryQueue
+{
+    private final int maxQueuedQueries;
+    private final AtomicInteger queryQueueSize = new AtomicInteger();
+    private final AsyncSemaphore<QueueEntry> asyncSemaphore;
+
+    QueryQueue(Executor queryExecutor, SqlQueryManagerStats stats, int maxQueuedQueries, int maxConcurrentQueries)
+    {
+        checkNotNull(queryExecutor, "queryExecutor is null");
+        checkNotNull(stats, "stats is null");
+        checkArgument(maxQueuedQueries > 0, "maxQueuedQueries must be greater than zero");
+        checkArgument(maxConcurrentQueries > 0, "maxConcurrentQueries must be greater than zero");
+
+        this.maxQueuedQueries = maxQueuedQueries;
+        this.asyncSemaphore = new AsyncSemaphore<>(maxConcurrentQueries,
+                queryExecutor,
+                queueEntry -> {
+                    QueryExecution queryExecution = queueEntry.dequeue();
+                    if (queryExecution == null) {
+                        // Entry was dequeued earlier and so this query is already done
+                        return Futures.immediateFuture(null);
+                    }
+                    else {
+                        SettableFuture<?> settableFuture = SettableFuture.create();
```

Comment:
I think this section is getting too intertwined with the logic of the queue management. I think we can abstract this out a bit by just having the QueryQueue take an object T that encapsulates the QueryExecution and nextQueues. And exposes 2 public functions: getCompletionFuture() (returning a ListenableFuture) and start(). I think it might be a little cleaner here.